provider/pkg/await/await.go

```diff
@@ -213,6 +214,37 @@ func Creation(c CreateConfig) (*unstructured.Unstructured, error) {
 				outputs, err = client.Patch(
 					c.Context, c.Inputs.GetName(), types.ApplyPatchType, objYAML, options)
 
+				// Handle the preview scenario where we need to re-create the object due to immutable fields.
+				// To avoid the immutable field error reported by the API server, we append "-preview" to the name
+				// so that the dry-run operation can succeed. This is a workaround for the limitation of the
+				// server-side apply API, which does not support replacement dry-run operations.
+				//
+				// Note that the AlreadyExists error does not occur since we always send a PATCH request during
+				// creates with SSA enabled. This means that we currently always upsert the object.
+				if c.Preview &&
+					(apierrors.IsInvalid(err) && strings.Contains(err.Error(), apivalidation.FieldImmutableErrorMsg)) {
+					const previewSuffix = "-preview"
+					objName := c.Inputs.GetName()
+
+					// Set the name to have the "-preview" suffix. Ensure we don't have a string over 253 characters, which is
+					// the maximum length for a name in Kubernetes.
+					var previewName string
+					if len(objName) > 253-len(previewSuffix) {
+						previewName = objName[:253-len(previewSuffix)] + previewSuffix
```

Comment:
Should the string be converted to runes before truncating it?  See https://stackoverflow.com/a/74700627/3026310

Reply:
Updated to use [upstream K8's name generator](https://github.com/jpbetz/kubernetes/blob/4b5ee802e847a3ea8243946aeae588906bd6f79f/staging/src/k8s.io/apiserver/pkg/storage/names/generate.go#L49-L54) instead.